In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('./Data/popular_git_repositories.csv')
df.head()

Name   
0                 freeCodeCamp  \
1       free-programming-books   
2                      awesome   
3                      996.ICU   
4  coding-interview-university   

                                         Description   
0  freeCodeCamp.org's open-source codebase and cu...  \
1         :books: Freely available programming books   
2  😎 Awesome lists about all kinds of interesting...   
3  Repo for counting stars and contributing. Pres...   
4  A complete computer science study plan to beco...   

                                                 URL            Created At   
0       https://github.com/freeCodeCamp/freeCodeCamp  2014-12-24T17:49:19Z  \
1  https://github.com/EbookFoundation/free-progra...  2013-10-11T06:50:37Z   
2            https://github.com/sindresorhus/awesome  2014-07-11T13:42:37Z   
3                  https://github.com/996icu/996.ICU  2019-03-26T07:31:14Z   
4  https://github.com/jwasham/coding-interview-un...  2016-06-06T02:34:12Z   

             Updated At                                           Homepage   
0  2023-09-21T11:32:33Z                http://contribute.freecodecamp.org/  \
1  2023-09-21T11:09:25Z  https://ebookfoundation.github.io/free-program...   
2  2023-09-21T11:18:22Z                                                NaN   
3  2023-09-21T08:09:01Z                                    https://996.icu   
4  2023-09-21T10:54:48Z                                                NaN   

     Size   Stars  Forks  Issues  ...  Has Issues Has Projects Has Downloads   
0  387451  374074  33599     248  ...        True         True          True  \
1   17087  298393  57194      46  ...        True        False          True   
2    1441  269997  26485      61  ...        True        False          True   
3  187799  267901  21497   16712  ...       False        False          True   
4   20998  265161  69434      56  ...        True        False          True   

  Has Wiki  Has Pages  Has Discussions  Is Fork  Is Archived  Is Template   
0    False       True            False    False        False        False  \
1    False       True            False    False        False        False   
2    False       True            False    False        False        False   
3    False      False            False    False         True        False   
4    False      False            False    False        False        False   

   Default Branch  
0            main  
1            main  
2            main  
3          master  
4            main  

[5 rows x 24 columns]

## Are there any duplicated values?

In [3]:
print("Number of duplicated values:", df.duplicated().sum())

Number of duplicated values: 0


## With each numerical column, how are values distributed?
- What is the percentage of missing values?
- Min? max? Are they abnormal?

### Percentage of missing values

In [4]:
missing_vals = df.select_dtypes(include='number').isna().sum()
missing_percentage = missing_vals / len(df)
missing_percentage

Size        0.0
Stars       0.0
Forks       0.0
Issues      0.0
Watchers    0.0
dtype: float64

### Min? Max? Are they abnormal?

In [5]:
df.describe()

Size          Stars          Forks         Issues   
count  2.150290e+05  215029.000000  215029.000000  215029.000000  \
mean   5.428270e+04    1115.085142     234.207637      37.925178   
std    7.023978e+05    3992.372050    1242.967451     196.508610   
min    0.000000e+00     167.000000       0.000000       0.000000   
25%    3.780000e+02     237.000000      39.000000       3.000000   
50%    2.389000e+03     377.000000      79.000000      10.000000   
75%    1.528200e+04     797.000000     174.000000      28.000000   
max    1.050786e+08  374074.000000  243339.000000   26543.000000   

            Watchers  
count  215029.000000  
mean     1115.085142  
std      3992.372050  
min       167.000000  
25%       237.000000  
50%       377.000000  
75%       797.000000  
max    374074.000000

For the 4 features **Stars, Forks, Issues and Watchers**, the *max* value seems to be very big compared to the rest of its column's data, 100 times larger than the 75% quantile.

But, this is **understandable** because it means there are some repositories that are *hugely more popular* than most other repositories.

## With each categorical column, how are values distributed?
- What is the percentage of missing values?
- How many different values? Show a few
- Are they abnormal?

### What is the percentage of missing values?

In [6]:
missing_vals = df.select_dtypes(exclude='number').isna().sum()
missing_percentage = missing_vals / len(df)
missing_percentage

Name               0.000009
Description        0.037353
URL                0.000000
Created At         0.000000
Updated At         0.000000
Homepage           0.635445
Language           0.074762
License            0.246660
Topics             0.000000
Has Issues         0.000000
Has Projects       0.000000
Has Downloads      0.000000
Has Wiki           0.000000
Has Pages          0.000000
Has Discussions    0.000000
Is Fork            0.000000
Is Archived        0.000000
Is Template        0.000000
Default Branch     0.000000
dtype: float64

Features with missing values are **Description, Homepage, Language and License.**

Notably, **Homepage** has a missing percentage of **0.635445** (over half of the data).

### How many different values? Show a few

In [7]:
categorical_vals = df.select_dtypes(exclude='number')

unique_vals = categorical_vals.apply(pd.Series.unique, axis=0)
unique_counts = unique_vals.apply(len)

Number of unique values for each column

In [8]:
print(unique_counts)

Name               196821
Description        206110
URL                215029
Created At         214922
Updated At         193011
Homepage            74198
Language              370
License                46
Topics             110123
Has Issues              2
Has Projects            2
Has Downloads           2
Has Wiki                2
Has Pages               2
Has Discussions         2
Is Fork                 1
Is Archived             2
Is Template             2
Default Branch       2326
dtype: int64


Some values for each column

In [9]:
print(unique_vals)

Name               [freeCodeCamp, free-programming-books, awesome...
Description        [freeCodeCamp.org's open-source codebase and c...
URL                [https://github.com/freeCodeCamp/freeCodeCamp,...
Created At         [2014-12-24T17:49:19Z, 2013-10-11T06:50:37Z, 2...
Updated At         [2023-09-21T11:32:33Z, 2023-09-21T11:09:25Z, 2...
Homepage           [http://contribute.freecodecamp.org/, https://...
Language           [TypeScript, nan, Python, JavaScript, C++, She...
License            [BSD-3-Clause, CC-BY-4.0, CC0-1.0, NOASSERTION...
Topics             [['careers', 'certification', 'community', 'cu...
Has Issues                                             [True, False]
Has Projects                                           [True, False]
Has Downloads                                          [True, False]
Has Wiki                                               [False, True]
Has Pages                                              [True, False]
Has Discussions                   

### Are they abnormal?

- **Is Fork** has only 1 value: False. We can remove this feature.

In [10]:
df = df.drop('Is Fork', axis=1)